src/components/sets/sets.js

```diff
@@ -0,0 +1,480 @@
+    const dropNode = this.findNode(dropKey);
+    if (dragNode.isCurrentSet || dropNode.isCurrentSet) {
+      return;
+    }
+    const loop = (data, key, callback) => {
+      data.forEach((item, index, arr) => {
+        if (item.setKey === key) {
+          callback(item, index, arr);
+        } else if (item.children) {
+          loop(item.children, key, callback);
+        }
+      });
+    };
+    const data = [...tabRoot.children];
+
+    // Find dragObject.
+    let dragObj;
+    loop(data, dragKey, (item, index, arr) => {
+      arr.splice(index, 1);
+      dragObj = item;
+    });
+
+    if (!dropToGap) {
+      // Set as last child of the dropKey node.
+      loop(data, dropKey, (item) => {
+        item.setChildren(item.children || []);
+        item.setChildren([...item.children, dragObj]);
+      });
+    } else if (insertBottom) {
+      // Set as first child of the dropKey node.
+      loop(data, dropKey, (item) => {
+        item.setChildren(item.children || []);
+        const newChildren = [...item.children];
+        newChildren.unshift(dragObj);
+        item.setChildren(newChildren);
+      });
+    } else {
+      // Set at intermediate position within dropKey node's children.
+      let ar;
+      let i;
+      loop(data, dropKey, (item, index, arr) => {
+        ar = arr;
+        i = index;
+      });
+      if (dropPosition === -1) {
+        ar.splice(i, 0, dragObj);
+      } else {
+        ar.splice(i + 1, 0, dragObj);
+      }
+    }
+
+    tabRoot.setChildren(data);
+    tabRoot.updateChildKeys();
+    this.emitTreeUpdate();
+  }
+
+  /**
+   * Set isEditing to true for a node of interest.
+   * @param {string} setKey The key of the node of interest.
+   */
+  startEditing(setKey) {
+    const node = this.findNode(setKey);
+    node.setIsEditing(true);
+    this.emitTreeUpdate();
+  }
+
+  /**
+   * Delete a node of interest.
+   * @param {string} setKey The key of the node of interest.
+   */
+  deleteNode(setKey) {
+    const parentNode = this.findParentNode(setKey);
+    if (!parentNode) {
+      return;
+    }
+    const nodeIndex = parentNode.children.find(c => c.setKey === setKey);
+    if (!nodeIndex) {
+      return;
+    }
+    parentNode.children.splice(nodeIndex, 1);
+    this.emitTreeUpdate();
+  }
+
+  /**
+   * Change a node's name.
+   * @param {string} setKey The key of the node of interest.
+   * @param {string} newName The new name value to assign.
+   * @param {boolean} stopEditing Whether to also set isEditing to false.
+   */
+  changeNodeName(setKey, newName, stopEditing) {
+    const node = this.findNode(setKey);
+    node.setName(newName);
+
+    if (stopEditing) {
+      node.setIsEditing(false);
+      if (node.wasPreviousCurrentSet) {
+        node.setWasPreviousCurrentSet(false);
+      }
+    }
+    if (node.isCurrentSet) {
+      node.setIsCurrentSet(false);
+    }
+    // TODO: check for existence of duplicate keys before setting the key.
```

Comment:
A potential place it could happen down the line: 
- You load factors which have pre-defined keys from the server. 
- You do things to the tree.
- You export the tree, lets say tree A.
- You refresh the page.
- Factors are loaded again.
- You do things to the tree, lets say tree B.
- You import tree A.
- The tree A's root becomes a child of tree B's root.
- You drag a factor node with key x from tree A into the factors of tree B, so now there are two nodes with the same key x under tree B's factors node.

Reply:
Or, even without import/export, the tree of keys loaded from the server looks like
- all
- all.my-factors
- all.my-factors.neurons
- all.your-factors
- all.your-factors.neurons

You drag all.your-factors.neurons under all.my-factors
The path-updating procedure keeps the "key tail", and now the tree looks like
- all
- all.my-factors
- all.my-factors.neurons
- all.my-factors.neurons
- all.your-factors